tools/drivewatch.go

```diff
@@ -575,6 +579,122 @@ func (dw *driveWatch) needRingUpdate(rd ringData) bool {
 	return false
 }
 
+func (dw *driveWatch) checkMissingDispersionObjects(rd ringData, cancel chan struct{}) error {
+	policy := rd.p.Index
+	db, err := dw.getDbAndLock()
+	defer dw.dbl.Unlock()
+	if err != nil {
+		return err
+	}
+	tx, err := db.Begin()
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+	rows, err := tx.Query(`
+	SELECT partition, account, container, object from dispersion_report_detail
+	WHERE dispersion_report_id = (
+		SELECT id FROM dispersion_report
+		WHERE policy = ? AND rtype = ? ORDER BY create_date DESC LIMIT 1)
+	`, policy, "object")
+	if err != nil {
+		dw.logger.Error("SELECT for max checkMissingDispersionObjects", zap.Error(err))
+		return err
+	}
+	defer rows.Close()
+	missingPartitions := map[int64]string{}
+	for rows.Next() {
+		var p int64
+		var a, c, o string
+		if err := rows.Scan(&p, &a, &c, &o); err == nil {
+			missingPartitions[p] = fmt.Sprintf("%s/%s/%s",
+				common.Urlencode(a), common.Urlencode(c), common.Urlencode(o))
+		} else {
+			dw.logger.Error("SELECT for checkMissingDispersion", zap.Error(err))
+		}
+	}
+	longMissingPartitions := []int64{}
+
+	for p := range missingPartitions {
+
+		rows, err := tx.Query(`
+			SELECT drd.partition
+			FROM dispersion_report dr LEFT JOIN dispersion_report_detail drd
+			ON dr.id = drd.dispersion_report_id AND drd.partition = ?
+			WHERE dr.policy = ? AND dr.rtype = ? AND dr.create_date > ?
+			`, p, policy, "object", time.Now().Add(-dw.rescueMissingPartAge))
+
+		if err != nil {
+			dw.logger.Error("SELECT for reports checkMissingDispersionObjects", zap.Error(err))
+			return err
+		}
+		defer rows.Close()
+		isReportWithPart := false
+		for rows.Next() {
+			var p sql.NullInt64
```

Comment:
Shadowing `p` here is confusing. Also, should this code be checking p.Valid somewhere?

Reply:
change p above to part, changed to Valid because am really checking for null